plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java

```diff
@@ -197,12 +197,10 @@ public Stream<DeltaLakeTransactionLogEntry> getCheckpointTransactionLogEntries(
             checkState(metadataAndProtocol.isPresent(), "metadata and protocol information is needed to process the add log entries");
         }
 
-        Stream<DeltaLakeTransactionLogEntry> resultStream = Stream.empty();
-        for (Location checkpointPath : getCheckpointPartPaths(checkpoint)) {
-            TrinoInputFile checkpointFile = fileSystem.newInputFile(checkpointPath);
-            resultStream = Stream.concat(
-                    resultStream,
-                    stream(getCheckpointTransactionLogEntries(
+        return getCheckpointPartPaths(checkpoint).stream()
+                .map(fileSystem::newInputFile)
+                .flatMap(checkpointFile -> {
+                    CheckpointEntryIterator checkpointEntryIterator = getCheckpointTransactionLogEntries(
```

Comment:
Could there be a potential resource leak here (which might not be introduced as part of this PR)? We call `TrinoInputFile.newInput()` within `ParquetPageSourceFactory.createPageSource` as part of the constructor of `CheckpointEntryIterator`. If some later part of the constructor fails, we likely won't get to close the file.

Reply:
> We call TrinoInputFile.newInput() ...

I'm assuming you are refferring to https://github.com/trinodb/trino/blob/875fe1b8a891e3293f0baf48cb4eea7b815b8b3d/plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/TrinoParquetDataSource.java#L39-L47


> If some later part of the constructor fails, we likely won't get to close the file.

There is actually exception handling in place for this specific scenario. See
https://github.com/trinodb/trino/blob/875fe1b8a891e3293f0baf48cb4eea7b815b8b3d/plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java#L285-L290


I'm adding though exception handling in the `CheckpointEntryIterator` constructor to make sure that the page source gets closed in case that there are eventual exceptions happening after initializing the `pageSource` field.